machineOSBuild and machineOSImage API

cdoern opened the pull request:
machineOSBuild and Image are meant to represent the two components of an on cluster build, making this commit the user-facing API for on cluster builds. 

MOSBuilds comprise of the actual processes of kicking off an monitoring a build. A build also contains user options like the push/pull spec for registries as well as the custom content in a containerfile. The state of a build will be tracked in here as well, and users will be able to view this in printer columns. A build is either: "Built", "BuildPrepared", "Building", "BuildFailed", "BuildInterrupted", or "BuildRestarted"

 a MOSImage is the result of a build. the Image objects are intended to be trackable and easily related to an MCP. The goal here is to remove much of the split state reporting for on cluster builds and consolidate where the user sets config options. These images have status conditions as well related to their Age and their Rollout to the nodes on a pool. The valid conditions are "RollingOut", "RolledOut", "RolloutFailed". There is also an "Age" printer column detailing the age of the image.

cdoern commented:
/jira refresh

cheesesashimi reviewed (changes requested):
Overall, this is a great first pass! I do have a lot of questions and thoughts about things. I'm happy to discuss them further if anything is unclear.

inesqyx commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+type MachineOSBuildSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineOSObjectReference `json:"machineConfigPool"`
```
I agree with what @cheesesashimi has suggested here. Doing so will make the tracking of targted MCP, the renderedMC and the renderedMachineOSImage built from the renderedMC clearer. Answers the question below as well.

JoelSpeed replied:
Bonus points, you can tailor the godoc and validations on the `name` field in each to be specific to the use case, and, if you expand a new field in the future, you only affect one specific use case that needs the new field

inesqyx commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+	// I am thinking we keep track of their name and why they failed
+	// +listType=set
+	// +optional
+	RelatedBuilds []PriorMachineOSBuilds `json:"relatedBuilds,omitempty"`
```
thought: I feel that we could find a more intuitive name for `RelatedBuilds`, e.g. `BuildHistory`, to suggest it is a list of old iterations.

cdoern replied:
sure, yeah. That makes more sense.

inesqyx commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+	FinalImagePushSecret string `json:"finalImagePushSecret"`
+	// finalImagePullSecret is the secret used to pull the base image.
+	// +kubebuilder:validation:Required
+	FinalImagePullSecret string `json:"finalImagePullSecret"`
```
question: shall the secret be named as `BaseImagePullSecret` if it is the secret used to pull the base image?

cheesesashimi replied:
There should be, at least for now. I feel like we might be conflating `MachineOSBuild` and `MachineOSBuildConfig` (although the latter is outside the scope of this PR).

cdoern replied:
I think if we are making these objects exist for the lifetime of a pool they are associated with, it makes sense to fold most (if not all) config and settings into the MachineOSBuild object @cheesesashimi  wdyt?

inesqyx commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+	// finalImagePullSpec describes the location of the final image.
+	// +kubebuilder:validation:Required
+	FinalImagePullSpec string `json:"finalImagePullSpec"`
+}
```
question: by the configmap method, we also allow the user to the pass in `ImageBuilderType`, but there is no similar field for the MachineOSBuild API as far as I have observed. Are we still supporting input for it?

cdoern replied:
ah! yeah I forgot that :)

JoelSpeed commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+	// renderedMachineOSImage describes the machineOsImage object created to track image specific information
+	// machineConfig is a reference to the MC that was used to build this image.
+	// +optional
+	RenderedMachineOSImage MachineOSObjectReference `json:"renderedMachineOSImage"`
```
Who is expected to set this? If it's a controller, then, status instead maybe?

JoelSpeed commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+	BuildInputs BuildInputs `json:"buildInputs"`
+	// currentConfig is the currently running config on the MCP
+	// +optional
+	CurrentConfig string `json:"currentConfig"`
```
Good question, I don't have as much background but we should make sure there's a clear source of truth for when a build needs to be triggered, and this delta between current and desired sounds like it needs to be calculated by a controller, not a user, so wouldn't expect these in spec.

cdoern replied:
desired will be necessary. We need something to trigger the build. Similarly to how currently we query the MCP and look at desired/current to trigger a build, something on the MachineOSBuild will need to do that now. Otherwise we will need sync loops in the MachineOSBuilder for MCPs and MOSB, I'd like to remove the MCP sync loop if possible.

JoelSpeed replied:
What and who would set the desired? Is the desired likely to change frequently?

cheesesashimi replied:
question: Shouldn't we make this a `RenderedMachineConfigReference`? Also, do we still need the `CurrentConfig` / `DesiredConfig` distinction here? Maybe it should just be `RenderedMachineConfig`?

cdoern replied:
I am implementing this in MCO, we do need currentConfig for a few parts in the buildController. I can go dig them up if need be

JoelSpeed commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+type MachineOSBuildStatus struct {
+	// buildStart describes when this build began
+	// +kubebuilder:validation:Required
+	BuildStartTime metav1.Time `json:"buildStart"`
```
Use a pointer here even though it's required. The json marshaller will render this as the zero time value, which will pass the required validation, but is not what you want.

You *have* to use a pointer for an optional `metav1.Time` so end time will need to be a pointer too

JoelSpeed commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
```
Conditions by convention must be the first item in the status

cdoern replied:
moved them!

JoelSpeed commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+	// +optional
+	BuildEndTime metav1.Time `json:"buildEnd"`
+	// conditions are state related conditions for the build. Valid types are:
+	// BuildPrepared, Building, BuildFailed, BuildInterrupted, BuildRestarted, and Built
```
I wonder if you need the `Build` prefix on all of these, given the context of the object, WDYT?

Suggestion, maybe instead of `Built` you could use `Ready` which is similar to other APIs.

I'd be interested to learn more about the other condition types, I wonder if some may be better consolidated, interrupted and restarted perhaps are failure conditions of Building?

cheesesashimi replied:
I agree. I don't think the `Build` prefix is needed here.

JoelSpeed commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+	// I am thinking we keep track of their name and why they failed
+	// +listType=set
+	// +optional
+	RelatedBuilds []PriorMachineOSBuilds `json:"relatedBuilds,omitempty"`
```
Maybe I need to read the enhancement, but, perhaps take a look at the Job construct in K8s and see how they handle tracking history, if at all.

Would I not expect to be able to `oc get machineosbuild` and see the historical items?

cdoern replied:
this part is still being figured out. I am leaning towards this flow:

1. one MachineOSBuild item per pool or group of nodes using the same config
2. this object exists for the lifetime that rendered-pool-abc configs exist and need to be managed
3. this field exists for the purpose of tracking "Failed" or "Interrupted" builds

I was originally thinking each BUILD gets its own MachineOSBuild object. This could make sense as well? thoughts @cheesesashimi ?

JoelSpeed replied:
How does a pool opt in to an on cluster builds?

JoelSpeed commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+type BuildProgress string
+
+const (
+	MachineOSBuildPrepared    BuildProgress = "BuildPrepared"
+	MachineOSBuiling          BuildProgress = "Building"
+	MachineOSBuildFailed      BuildProgress = "BuildFailed"
+	MachineOSBuildInterrupted BuildProgress = "BuildInterrupted"
+	MachineOSBuildRestarted   BuildProgress = "BuildRestarted"
+	MachineOSBuilt            BuildProgress = "Built"
```
Would appreciate godoc explaining these and their use case in particular

JoelSpeed commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+type PriorMachineOSBuilds struct {
+	// name is the name of the build
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
```
Oh good question, I had assumed from where this is in spec, that it was another object, but if it's pods then tracking that is definitely like a Job

cdoern replied:
this tracks old MachineOSBuilds (they will still be around until they are rotten)

cheesesashimi replied:
Just to make sure I understand correctly:

We have a MachineOSBuild object. Whenever something changes, we're creating a child MachineOSBuild object? Kind of like what we're doing with MachineConfigs and rendered MachineConfigs? Or am I misunderstanding this?

I was under the impression that a MachineOSBuild object holds the needed configuration(s) for a pool to do a build. Whenever certain elements of the configuration change, the MachineOSBuilder controller detects that change and then launches a build. This would then produce a MachineOSImage (which I know is beyond the scope of this PR). And whenever a new MachineOSImage is produced, that gets assigned to a MachineConfigPool so that we can roll it out.

JoelSpeed replied:
Not a fan of child objects (sorry MachineConfig pattern), but yes, I was on Zack's understanding as well

cheesesashimi replied:
Yeah, I think that's the best way forward.

JoelSpeed commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+type BuildInputs struct {
+	// osImageURL is the base OSImage we use to build our custom image.
+	// +kubebuilder:validation:Required
+	OSImageURL string `json:"osImageURL"`
```
Is this a URL? Can we use CEL URL validation on it?

cheesesashimi replied:
This would be an image pullspec. Similarly, is there a validator for image pullspecs? Or is this something we'd have to validate ourselves?

JoelSpeed commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+	OSImageURL string `json:"osImageURL"`
+	// finalImagePushSecret is the secret used to connect to a user registry.
+	// +kubebuilder:validation:Required
+	FinalImagePushSecret string `json:"finalImagePushSecret"`
```
Is this a secret reference? Or a literal secret?

cdoern replied:
reference. I made a new object reference to track these

JoelSpeed commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+	FinalImagePushSecret string `json:"finalImagePushSecret"`
+	// finalImagePullSecret is the secret used to pull the base image.
+	// +kubebuilder:validation:Required
+	FinalImagePullSecret string `json:"finalImagePullSecret"`
```
Object reference yes, but don't re-use the corev1 version, create your own

cdoern replied:
yeah, object references would be nice here.

JoelSpeed commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+	FinalImagePullSecret string `json:"finalImagePullSecret"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// +optional
+	ContainerFile []byte `json:"containerFile"`
```
> quibble: I think Containerfile may be preferred instead of ContainerFile.

In this context, treat me like a 4 year old, why would it be `Containerfile`? Is this some special keyword?

cheesesashimi replied:
@JoelSpeed I'm just used to it being referred to in the Podman docs as `Containerfile` instead of styled as `ContainerFile`. It's mostly just for the sake of consistentcy.

cheesesashimi replied:
@cdoern I think string would be more ergonomic since a cluster admin could just do something like: `$ oc get machineosbuild/worker -o yaml` and be able to see the Containerfile field without having to encode / decode it.

JoelSpeed replied:
Would be good to see if we have other APIs that refer to either a `Dockerfile/DockerFile` or a `ContianerFile/`Containerfile` for comparison. I'm 50/50 on this one, since it does appear to be a keyword, it's not a Container file per say, but a Containerfile, so it may make sense to stick with `containerfile`

cdoern commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+	// renderedMachineOSImage describes the machineOsImage object created to track image specific information
+	// machineConfig is a reference to the MC that was used to build this image.
+	// +optional
+	RenderedMachineOSImage MachineOSObjectReference `json:"renderedMachineOSImage"`
```
refers to the `MachineOSImage` (separate PR) object created by this build

cdoern commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+// BuildInputs holds all of the information needed to trigger a build
+type BuildInputs struct {
```
thanks :)

cheesesashimi commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,174 @@
+	BaseOSImageURL string `json:"baseOSImageURL"`
+	// baseImagePullSecret is the secret used to pull the base image.
+	// +kubebuilder :validation:Required
+	BaseImagePullSecret string `json:"baseImagePullSecret"`
```
thought: I'm not opposed to keeping it around, but I'm not sure that we still need the `BaseImagePullSecret` field here anymore. I think the preferred way forward will be using the global pull secret instead since it should be able to pull the base OS image.

JoelSpeed replied:
Curious why the timeline comes into this decision? What would removing this have to do with the timeline?

JoelSpeed replied:
Ok so, perhaps there should be a spec field to allow the user to override, but a status field to show the actual secret that was used?

> don't think this logic of auto-retrieval will be done for 4.16

It sounds like you're thinking to ship this feature in parts, are there clear separations of the feature's within this that can and cannot be completed in time? Is there a way to make it clear what is, and is not supported for 4.16?

cheesesashimi commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,174 @@
+const (
+	OCPBuilder     MachineOSImageBuilderType = "OpenShiftImageBuilder"
+	PodBuilder     MachineOSImageBuilderType = "PodImageBuilder"
+	DefaultBuilder MachineOSImageBuilderType = "Default"
```
question: I presume we'll make the link between the `DefaultBuilder` and `OCPBuilder` within the MCO?

cdoern replied:
yep!

cheesesashimi commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,174 @@
+	// I am thinking we keep track of their name and why they failed
+	// +listType=set
+	// +optional
+	BuildHistory []PriorMachineOSBuilds `json:"buildHistory,omitempty"`
```
thought: I'm wondering if we want to somehow link the underlying build pods / ConfigMaps / etc. to this object somehow. Within BuildController, we could do that with labels and the like, but I wonder if it makes sense to link them here as well.

(I'm not asking you to add them here right now though. That could be a future addition.)

cdoern replied:
I think I actually will add these now, if its too much, I can remove it. I will add these in my next commit

cheesesashimi commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,174 @@
+	FinalImagePullSecret ImageSecretObjectReference `json:"finalImagePullSecret"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// +optional
+	Containerfile []byte `json:"containerFile"`
```
thought: We still need to figure out if making this a `[]byte` would require cluster admins to base64 encode / decode their Containerfiles.

cdoern replied:
i meant to switch this to string, will do that now

cheesesashimi commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,174 @@
+type BuildInputs struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// Valid options are: OpenShiftImageBuilder, PodImageBuilder, and Default (OpenShiftImageBuilder)
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
```
note: Right now, I don't think there is a way to bind an image builder type to a specific pool. This feels like it should be a global option in the future.

cdoern replied:
this can be MCO validation that we run I believe.

cdoern commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,218 @@
+	BuildInputs BuildInputs `json:"buildInputs"`
+	// currentConfig is the currently running config on the MCP
+	// +optional
+	CurrentConfig RenderedMachineConfigReference `json:"currentConfig"`
```
maybe we don't need this

cdoern commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,218 @@
+	FinalImagePullSecret ImageSecretObjectReference `json:"finalImagePullSecret"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// +optional
+	Containerfile string `json:"containerFile"`
```
make this a set

hexfusion replied:
if this is a set how many  Containerfiles can you have before you hit 1mb resource limits?

cheesesashimi replied:
OpenShift supports a total of four different architectures (amd64, aarch64, ppc64le, s390x). Would someone actually build a cluster with all four architectures in the same cluster? Probably not, but it sounds like something I would do just to say that I did it :laughing:

That said, I'd say a cluster admin would only have two different architectures within the same cluster. But would adding a max of 4 cause excessive resource usage?

cheesesashimi replied:
As an aside, I have a question about the implementation detail here: How would we assign a Containerfile to a particular architecture? Would it make sense to do something like this?

```golang
type ContainerfileArch string

const (
	// Or we import them from here? https://github.com/openshift/api/blob/master/samples/v1/types_config.go#L178-L191
	// The only downside of doing that is they are just strings as opposed to a separate type.
	AMD64ContainerfileArch ContainerfileArch = "amd64"
	AARCH64ContainerfileArch ContainerfileArch = "aarch64"
	// ...
)

type Containerfile struct {
	Arch ContainerfileArch `json:"arch"`
	Contents string `json:"contents"`
}

// Then BuildInputs becomes something like:
type BuildInputs struct {
	// ...
	Containerfiles []Containerfile `json:"containerfiles"`
	// ...
}
```

hexfusion commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,218 @@
+	BinaryData ConfigMapData `json:"binaryData"`
+}
+
+// ConfigMapData is used to map the data in the OCB configmaps so that they can be viewed in the object's statuss
```
```suggestion
// ConfigMapData is used to map the data in the OCB configmaps so that they can be viewed in the object's statuses
```

hexfusion commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,218 @@
+// Refers to the name of a (future) MachineOSImage (e.g., "worker-os-image-167651b10ec98af17971d6a47df9e22f", etc.):
+type MachineOSImageReference struct {
+	// name is the name of the MachineOSImage object attached to this MachineOSBuild.
+	// +kubebuilder:validation:Required
```
We are going to need some validation here  Min/Max/Regex?

hexfusion commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,218 @@
+	BaseOSImageURL string `json:"baseOSImageURL"`
+	// baseImagePullSecret is the secret used to pull the base image.
+	// +kubebuilder :validation:Required
+	BaseImagePullSecret ImageSecretObjectReference `json:"baseImagePullSecret"`
```
we talked about this in the review session but there really only is a single pull secret for the cluster so I am not sure what this supposed to be for.  Is this a path to a pullsecret on disk?

cdoern replied:
The reason we have this is because as of 4.16, the user will need to specify their own base image pull secret, we do not have the plumbing to get this for them.

The plan in 4.17+ is to add some sort of validation so they cannot edit this? Though @cheesesashimi can you think of a scenario where we want to allow them to specify a custom base image pull secret?

cheesesashimi replied:
I could see them potentially asking for that in a Hypershift context. @yuqi-zhang In Hypershift clusters, is it common for the guest clusters to use the same global pull secret as the control-plane cluster? Is there a situation where it would be desirable for someone to do that?

hexfusion commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,218 @@
+	BaseOSExtensionsImageURL string `json:"baseOSExtensionsImageURL"`
+	// baseOSImageURL is the base OSImage we use to build our custom image.
+	// +kubebuilder:validation:Required
+	BaseOSImageURL string `json:"baseOSImageURL"`
```
You will want to add validation for this url

hexfusion commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,218 @@
+	StartTime *metav1.Time `json:"startTime"`
+	// endTime describes when the build ended
+	// +optional
+	EndTime *metav1.Time `json:"endTime"`
```
Would adding metrics be more useful for these durations ?

cdoern replied:
The goal here is that if a build is taking 10hrs, the user can restart it via some mechanism. I guess this would be deleting and re-creating the whole object? or maybe we can build in some sort of restart annotation to this object for a forced restart?

JoelSpeed replied:
Start and end is consistent with other objects we've seen that perform one time actions

Typically the idiom for restarting, would be to delete and recreate, which is another reason this API should probably be some sort of config which creates a child build for each attempt at building (not sure if that's the shape of this but it didn't look like it when I initially scanned)

hexfusion commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,218 @@
+type ImageSecretObjectReference struct {
+	// name is the name of the secret used to push or pull this MachineOSBuild object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
```
Is the user expected to create these references or a controller?

cdoern replied:
User (for now :) ). Until we add code to get these secrets for the user they are going to populate the CR will all of their config options. This is what they currently do in the "configmap" version of On-Cluster-Layering, so not much of a change here.

cdoern replied:
when the controller fails. As long as the actual formatting of these secrets object references are valid, the build controller will be the one actually using the secrets and will fail if they are inaccurate.

hexfusion commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,218 @@
+type BuildInputs struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// Valid options are: OpenShiftImageBuilder, PodImageBuilder, and Default (OpenShiftImageBuilder)
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
```
Since this is user-generated how do you plan on gating the available builder types?

cdoern replied:
User can edit this, but we will verify they are using one of the 3 constants via the MCO. Unless there is a kubebuilder validation we can run to get the same effect?

JoelSpeed replied:
```
+kubebuilder:validation:Enum:=A;B;C
```
Validate as much as you can at the API level, use this^ for an enum

hexfusion commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,218 @@
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+	// baseOSExtensionsImageURL is the base Extensions image used in the build process
+	// +kubebuilder:validation:Required
+	BaseOSExtensionsImageURL string `json:"baseOSExtensionsImageURL"`
```
please add URL validations

hexfusion reviewed (commented):
First pass will take a deeper look soon. I am concerned with the possible size of some of these values. We should make efforts to calculate what is possible in a single CR and enforce that on the API level.

cdoern commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,218 @@
+	Name string `json:"name"`
+	// Data is the string data in the configmap
+	// +optional
+	Data ConfigMapData `json:"data"`
```
Configmap data will always been small in this case, sorry I should specify this. @cheesesashimi do you have a rough estimate of how many entries are in these MCO namespaced OCB configmaps?

These are just used to track temp data for the build.

cheesesashimi replied:
To give a bit of context behind what these ConfigMaps are and why they're used:

These ConfigMaps contain the rendered Containerfile and the rendered MachineConfigs as build inputs. The Containerfile is reasonably small, but a MachineConfig might be a bit larger. In some cases, it is not unheard of for a rendered MachineConfig to be greater than 1 MB in size (once serialized to JSON). The reason I know this is because this became a concern for Hypershift, which stuffs rendered MachineConfigs into ConfigMaps and we had to implement gzip / base-64 encoding to work around the 1 MB size limit that ConfigMaps have.

We do a similar thing in the BuildController to avoid giving the build pods elevated permissions. We stuff the rendered Containerfile and rendered MachineConfigs into ConfigMaps and mount them into the build context. These exist for the duration of the build. Assuming the build is successful, these objects are deleted. If the build is unsuccessful, the objects remain so that a cluster admin can troubleshoot.

Doing some back-of-napkin math here, let's assume the worst-case that we have an OpenShift cluster with two MachineConfigPools, one of which has both amd64 and aarch64 nodes:

Control Plane Pool:
- Rendered MachineConfig (1 MB max size)
- Rendered Containerfile (1 MB max size)
2 MB total, worst-case.

Worker Pool:
- Rendered MachineConfig (1 MB max size) - There is no way to set an architecture-specific MachineConfig, so this ConfigMap could be shared between both the amd64 and aarch64 builds.
- Rendered AMD64 Containerfile (1 MB max size)
- Rendered ARM64 Containerfile (1 MB max size) - We could dedupe these if we've determined that the Containerfiles are equal.
3 MB total, worst-case.

Cluster total: 5 MB, worst-case. This is assuming that one is running all of the builds concurrently. Assuming no builds are running, these ConfigMaps will not exist. Each additional pool would be at minimum, 2 MB for homogenous clusters, with 1 MB additional for each additional architecture supported.

I hope this helps to answer the question. My initial thought is that maybe we shouldn't inline this data. But I could be convinced otherwise.

cheesesashimi replied:
For a more realistic example, on a freshly-provisioned OpenShift cluster with on-cluster builds enabled, I pulled down the ephemeral ConfigMaps to get an idea of the size of their contents:

```console
# This is the size of the rendered MachineConfig for this pool once it's been gzipped / base64-encoded, and stuffed into a ConfigMap.
$ oc get configmap/mc-rendered-layered-950b76f1b1fb344d902cb041ad15d695 -n openshift-machine-config-operator -o json | jq -c -r '.data["machineconfig.json.gz"]' | wc -c
   49825

# This is the size of the rendered Containerfile for this pool once it's been rendered and stuffed into a ConfigMap.
$ oc get configmap/dockerfile-rendered-layered-950b76f1b1fb344d902cb041ad15d695 -n openshift-machine-config-operator -o json | jq -c -r '.data.Dockerfile' | wc -c
    3327
```

These numbers are probably much closer to the average case.

hexfusion replied:
Thanks for the details my concern with inlining this data into the Spec would be with the size limit of the CR.  I believe you would be safer having references to resources but then you need to maintain them. Would limiting the size of these fields also limit future use cases?

cdoern commented:
@hexfusion thanks for the review. Most of your concerns about size I am going to address by enforcing a limit. The Containerfile one is tricky. Currently we only allow 1, but with @cheesesashimi 's point about multi-arch, I might allow 2? I think this'll be below 1mb still, and if not we can enforce 1.

I am going to work on adding validations today for these concerns and will let you know when its up to date.

cgwalters commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,218 @@
+	// failed indicates that during the build or preparation process, the build failed.
+	MachineOSBuildFailed BuildProgress = "Failed"
+	// interrupted indicates that the user stopped the build process by modifying part of the build config
+	MachineOSBuildInterrupted BuildProgress = "Interrupted"
```
I know there's a ton of comments here and sorry if this is duplicating existing feedback but at a high level a goal I think we should have here is that things related to the booted host container should "feel the same" as "app" containers - where it makes sense.

Most of the other states here apply equally well to application container images.  So why not just import and use e.g. https://github.com/openshift/api/blob/master/build/v1/consts.go#L153 ?

cdoern replied:
@cgwalters sure, I'd be fine with using some of those. I am not sure if there is a 1:1 match for some of the states (like InProgress, Restarted). But Interrupted could map well to Cancelled and Failed to the one you linked above. 

Would you be ok with picking and choosing the ones that match with some new custom states here?

cdoern replied:
We want users to be able to track some sort of "progress". Since we are giving them the ability to control their build inputs but we actually run the build, we don't want to create too much of a black box that leads to more MCO bugs if that makes sense? So states like `Building` and `BuildRestarted`, along with the rationale of what happened specific to MCO, is important

hexfusion replied:
Sorry to also comment inline here but seems possibly appropriate.

If the Controller reported a Degraded true status on build failure with a reasonable message. Couldn't debug be done via container logs and or assisted by events? That seems like a reasonable/existing pattern. Status isn't really a full logger just a high-level signal of where to look next? Then also does the user need information if things are good or only when they are not?

cdoern replied:
yeah, it is reasonable to assume they will debug into the container/pod.

Again, I do think that the user would benefit from this flow though given the fact that they have no idea what is going on in the build. 

With our MCP object (and a few others) we tend to update state in both good and bad scenarios, so I'd like to keep this object operating similarly if possible. 

I believe we can change these state constants in tech preview. So should we continue as is on the state reporting front and change as needed in future iterations so we don't get blocked on this?

cdoern replied:
We are opting to use these types for a few reasons

1. possible deprecation of the build API in the future ( we don't want to be dependent on those deprecated types)
2. Using custom types as opposed to re-using other people's types is better practice in the API where applicable.
3. not all of the types seem to fit for what we need.

Open to discussing this more, but for purposes of getting this into 4.16, this is what the team has discussed.

cc @mrunalp @hexfusion @cheesesashimi

cdoern commented:
no idea why these tests are failing, integration especially since that passes locally...

/retest-required

cheesesashimi commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,174 @@
+	BaseImagePullSecret string `json:"baseImagePullSecret"`
+	// finalImagePushSecret is the secret used to connect to a user registry.
+	// +kubebuilder:validation:Required
+	FinalImagePushSecret ImageSecretObjectReference `json:"finalImagePushSecret"`
```
That is what I meant. But if these can be set to a default value, that's fine.

cheesesashimi commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,218 @@
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	BuildHistory []BuildHistory `json:"buildHistory,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
```
I was thinking about a user-configurable `MaxRetries` field somewhere, but we can have that hard-coded to 3 times for now.

hexfusion replied:
I can't say I have strong feels just that you will want to try and dictate a clear understanding to the user of what to expect and then ensure you bound your limits to ensure capacity.

> My logic here is that is a build fails 3 times its most likely failing for the same or related reasons. We should then stop building so the Machine-OS-Builder doesn't do this forever.

Is this Unreconcilable in the eyes of your controller at 3 failures? Important to think of future use cases and consider your implemntation.

cdoern replied:
yeah, I would say this describes our Unreconcilable state. Maybe I will add a few things here 

1. a `Restart` knob to trigger a rebuild
2. a `MaxRetries` knob to let the user override the 3 default attempts to rebuild

hexfusion commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,237 @@
+	Data ConfigMapData `json:"data"`
+	// Binary data is the binary data in the configmap
+	// +optional
+	BinaryData ConfigMapData `json:"binaryData"`
```
same here

cdoern replied:
I will likely make this a reference

cdoern commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,237 @@
+type MachineOSBuildSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineConfigPoolReference `json:"machineConfigPool"`
```
the pool associated with this build object is crucial as it lets the build controller know what pool the image is for. This will then auto-populate the renderedMachineConfig reference and a few other entries in the object so that the build controller has the right inputs and rolls out the image to the right nodes.

Without the MCP ref, the build controller does not know who it is building for (or what it should build)

hexfusion replied:
OK good context, just thinking out loud but isn't build and deploy two separate actions? Is there an assumption that as soon as the build is completed it is deployed immediately?

hexfusion replied:
Would there be a case and maybe its covered in enhancement where user just builds image with the idea it could be deployed in the future. In that case could the existing mechanism for deploying images be used?

cdoern replied:
the image will be immediately deployed to the nodes. I don't know of a circumstance where it would be advantageous to batch build images. so for now, no separate deploy mechanism exists.

deads2k replied:
Separate actions here make sense to me.  I would expect whatever is telling nodes to "update to this level" to use the MachineOSConfig resource.

In which case, this should be a MachineOSConfig reference.

cdoern commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,237 @@
+}
+
+// Refers to the name of a rendered MachineConfig (e.g., "rendered-worker-ec40d2965ff81bce7cd7a7e82a680739", etc.):
+type RenderedMachineConfigReference struct {
```
This is used to store the machine config the build is targeting. I will add a doc here

cdoern commented:
/retest-required

cdoern commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,203 @@
+	// Valid options are: OpenShiftImageBuilder, PodImageBuilder, and Default (OpenShiftImageBuilder)
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+	// baseOSExtensionsImageURL is the base Extensions image used in the build process
+	// +kubebuilder:validation:Pattern=`^https://`
```
remove.

cdoern commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,203 @@
+
+const (
+	// RebuildNow indicates that the user wants to rebuild their image immediately
+	RebuildNow TriggerRebuild = "RebuildNow"
```
will this cause an infinite rebuild loop?

cdoern replied:
I will wire up logic in the MCO so that if he user sets `RebuildNow` or `RebuildDelayed`, the next rebuild will cause this to get get back to `NoRebuild`

hexfusion replied:
What is the mechanism for defining the delay interval for `RebuildDelayed`

cdoern replied:
Delayed will kick off a build when the current build (if there is one) either succeeds or fails. I will add a godoc detailing this

deads2k replied:
> I will wire up logic in the MCO so that if he user sets `RebuildNow` or `RebuildDelayed`, the next rebuild will cause this to get get back to `NoRebuild`

this is bad practice.  Fields should be owned by exactly one entity, not shared between users and controllers.  See the ForceRedeploymentReason mechanic employed by static pod operators: https://github.com/openshift/api/blob/master/operator/v1/types.go#L191

cdoern commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,203 @@
+	// +optional
+	FinalImagePullSecret ImageSecretObjectReference `json:"finalImagePullSecret"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// +listType=set
```
map -> key == arch, value == data

arch would be optional. both for regular clusters and multiarch

cdoern replied:
beef up godoc

cdoern commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,203 @@
+	// baseImagePullSecret is the secret used to pull the base image.
+	// +kubebuilder:validation:Required
+	BaseImagePullSecret ImageSecretObjectReference `json:"baseImagePullSecret"`
+	// finalImagePushSecret is the secret used to connect to a user registry.
```
godoc for push/pull secrets, why are they sep?

cdoern commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,203 @@
+	// +kubebuilder:validation:MinItems=0
+	// +kubebuilder:validation:MaxItems=2
+	// +optional
+	Containerfile []string `json:"containerFile"`
```
`ContainerfIleArch` custom type or something

hexfusion commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,247 @@
+	BuildPod BuildPodReference `json:"buildPod"`
+}
+
+// BuildPodReference refers to the pod used to build an image that either failed or was interrupted
```
why only failed builds?

cdoern replied:
failed, interrupted, etc. I guess we could put in progress or succeeded builds here too :) though, we will still limit it to the most recent 3 for size constraints

hexfusion commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,111 @@
+	MachineOSBuildInterrupted BuildProgress = "Interrupted"
+	// restarted indicates that this build has been either interrupted or failed, and now a new
+	// iteration of the process has begun.
+	MachineOSBuildRestarted BuildProgress = "Restarted"
```
could a build even if successfully built be forced to build again?

cdoern replied:
yeah. A successful build could happen using the wrong image builder. If the user changes the image builder mid-build and specifies "RebuildDelayed" then we will wait until this build executes completely.

deads2k replied:
Restart seems error prone.  Could we instead create a new instance of `machineosbuild` so that terminal states of machineosbuilds exist?

cdoern commented:
/test verify

cdoern commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,121 @@
+// +kubebuilder:printcolumn:name="Prepared",type="string",JSONPath=.status.conditions[?(@.type=="Prepared")].status
+// +kubebuilder:printcolumn:name="Building",type="string",JSONPath=.status.conditions[?(@.type=="Building")].status
+// +kubebuilder:printcolumn:name="Ready",type="string",JSONPath=.status.conditions[?(@.type=="Ready")].status
+// +kubebuilder:printcolumn:name="Interrupted",type="string",JSONPath=.status.conditions[?(@.type=="Interrupted")].status
```
`Interrupted` is distinct in that it is a user caused restart. We currently have an equivalent of this in the MCO, I think we need this or else we will be making our lives harder and the state reporting less clear.

`Ready` means the build has completed and is READY to deploy. open to changing this? I think all of these are flexible enough

cdoern commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,121 @@
+// +kubebuilder:printcolumn:name="Restarted",type="string",JSONPath=.status.conditions[?(@.type=="Restarted")].status
+// +kubebuilder:printcolumn:name="Failed",type="string",JSONPath=.status.conditions[?(@.type=="Failed")].status
+
+// MachineOSBuild describes a build process managed by the MCO
```
true. will update

cdoern commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,121 @@
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineOSBuildList describes all of the Builds on the system
```
nope, just builds. `MachineOSImage` (a future API) will be the deployment side of things

cdoern commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,121 @@
+	Items []MachineOSBuild `json:"items"`
+}
+
+// MachineOSBuildSpec describes user-configurable options as well as information about a build process.
```
this API isn't about deployment. So I don't think we should include too many of these refs. That was the `machineosimage` api that we canned and are going to introduce in 4.17

cdoern commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,121 @@
+	MachineOSReady BuildProgress = "Ready"
+)
+
+// Refers to the name of a (future) MachineOSImage (e.g., "worker-os-image-167651b10ec98af17971d6a47df9e22f", etc.):
```
sorry, I should have deleted this. MachineOSImage is not in scope here.

cdoern commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+// Refers to the name of a rendered MachineConfig (e.g., "rendered-worker-ec40d2965ff81bce7cd7a7e82a680739", etc.):
+// the build targets this MachineConfig, this is often used to tell us whether we need an update.
```
if the desired config changes on the pool, it will be represented here, and a build will be triggered. Will add a comment

cdoern commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,264 @@
+	Rebuild MachineOSRebuild `json:"rebuild"`
+}
+
+// MachineOSConfigStatus describes the status of builds associated with this MachineOSConfig
```
`// MachineOSConfigStatus describes the status this config object and relates it to the builds associated with this MachineOSConfig
`

cdoern commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,264 @@
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=3
+	// +optional
+	BuildHistory []BuildHistory `json:"buildHistory,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
```
build history, helpful to have info about why it failed (or succeeded), assoc. pods, configmaps etc.

deads2k replied:
> build history, helpful to have info about why it failed (or succeeded), assoc. pods, configmaps etc.

How about having Conditions instead and using those to present this status. Then leave build history to the machineosbuild + ordinal.

cdoern commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,264 @@
+	noArch ContainerFileArch = "noArch"
+)
+
+// Refers to the name of a MachineConfigPool (e.g., "worker", "infra", etc.):
```
build-controller. Will add a comment.

cdoern commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,264 @@
+	RebuildNow TriggerRebuild = "RebuildNow"
+	// RebuildDelayed indicates that the user wants to rebuild their image the next time the build controller is idle.
+	// if there is a build in progress, a rebuild will be triggered the next time that build fails or succeeds. This is often used
+	// when the user changes the top level configuration for a pool's build.
```
hmmm. It causes a new build. The final image pullspec will stay the same if the build failed.

cdoern replied:
and even if the build fails and the user is using a tag, the final img pullsped will stay the same

cdoern commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,264 @@
+}
+
+// BuildPodReference refers to the pod used to build an image that either failed or was interrupted
+type BuildPodReference struct {
```
this is used for status. I have is in the config file since the MOSC uses it in BuildHistory. Honestly, idk if we need this (or configmaps) @cheesesashimi voted to add those in, but as I am implementing in MCO, seems to be causing friction. Would be ok with removing this AND configmaps in the build history. @hexfusion

deads2k commented on machineconfiguration/v1alpha1/tests/machineosbuilds.machineconfiguration.openshift.io/OnClusterBuild.yaml:
```diff
@@ -0,0 +1,23 @@
+apiVersion: apiextensions.k8s.io/v1 # Hack because controller-gen complains if we don't have this
+name: "[TechPreview] MachineOSBuild"
+crdName: 0000_80_machine-config_01_machineosbuilds-TechPreviewNoUpgrade.crd.yaml
```
rdname should be the name of the CRD, not the filename.  machineosbuilds.machineconfiguration.openshfit.io I bet?

deads2k commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,124 @@
+	Items []MachineOSBuild `json:"items"`
+}
+
+// MachineOSBuildSpec describes user-configurable options as well as information about a build process.
```
are users actually expected to modify this?  I expected us to only want our controllers to create these.

cdoern replied:
true, only the controller will set this. I can update the godoc. Do you think it would be better if this lived in status? Are we allowed to have an object without anything in status?

deads2k commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,124 @@
+type MachineOSBuildStatus struct {
+	// version tracks the newest MachineOSBuild for each MachineOSConfig
+	// +kubebuilder:validation:Required
+	Version int64 `json:"version"`
```
must be positive

deads2k replied:
This seems like an input, not an output.  When an instance of machineosbuild is created, it's relative ordering is knowable and we don't want the thing consuming these to consume them in the wrong order.

cdoern replied:
so, are you saying this should live in the spec?

deads2k commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,124 @@
+	Version int64 `json:"version"`
+	// configGeneration tracks which version of MachineOSConfig this build is based off of
+	// +kubebuilder:validation:Required
+	ConfigGeneration int64 `json:"configGeneration"`
```
must be positive

deads2k commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,124 @@
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
```
conditions first item in status please

deads2k commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,124 @@
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// buildName is the build object reference associated with this MachineOSBuildObject
+	BuildName string `json:"buildName"`
```
We agreed to evolve away from Builds, right?  To reference an arbitrary thing, you want something closer to `RelatedObjects`, rigth?  https://github.com/openshift/api/blob/master/config/v1/types_cluster_operator.go#L66-L71

Bonus, if you use RelatedObject with a matching schema, an `oc adm inspect` will gather them.

If your controller is using this to track status of the thing building for it, you shoudl create a discriminated union that matches the ImageBuilderType.  If you later want to add different type (and I think you did), you'll want to start that discriminated union now.

cdoern replied:
yes, if we are moving away from builds, then this will probably reference a build pod. Though, something like RelatedObjects will fit the bill here.

deads2k commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,124 @@
+	// buildStart describes when the build started.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="buildStart is immutable once set"
+	// +kubebuilder:validation:Required
+	BuildStart *metav1.Time `json:"buildStart"`
+	// buildEnd describes when the build ended.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="buildEnd is immutable once set"
+	//+optional
+	BuildEnd *metav1.Time `json:"buildEnd"`
```
these can stay outside the discriminated union.

deads2k commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,124 @@
+	BuildEnd *metav1.Time `json:"buildEnd"`
+	// desiredConfig is the desired config we want to build an image for.
+	// +kubebuilder:validation:Required
+	DesiredConfig RenderedMachineConfigReference `json:"desiredConfig"`
```
not clear on how this can be status. this seems to be the input.

deads2k commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,124 @@
+	Version int64 `json:"version"`
+	// configGeneration tracks which version of MachineOSConfig this build is based off of
+	// +kubebuilder:validation:Required
+	ConfigGeneration int64 `json:"configGeneration"`
```
This seems like an input, not an output.  "MachineOSBuild for this level of MachineOSConfig" and then

1. if the MachineOSConfig I read from my cache is newer, permanently fail/skip this build.  A newer instance will appear.
2. if the MachineOSConfig I read from my cache is old wait until I see the one that should be used.

cdoern replied:
moved to spec

deads2k commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,124 @@
+	// finalImagePullSpec describes the fully qualified pullspec produced by this build that the final image can be. Must be in sha format.
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@').size() == 2 && self.split('@')[1].matches('^sha256:[a-f0-9]{64}$'))`,message="the OCI Image reference must end with a valid '@sha256:<digest>' suffix, where '<digest>' is 64 characters long"
+	// +optional
+	FinalImagePullspec string `json:"finalImagePullspec"`
```
from the perspective of the machineosbuild, this is a push spec, right?

cdoern replied:
yes. I was thinking of renaming it that.

deads2k commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,124 @@
+	// iteration of the process has begun.
+	MachineOSBuildRestarted BuildProgress = "Restarted"
+	// ready indicates that the build has completed and the image is ready to roll out.
+	MachineOSReady BuildProgress = "Ready"
```
So this is Complete?

cdoern replied:
yep, I can rename it to complete if that is the norm

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,233 @@
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@').size() == 2 && self.split('@')[1].matches('^sha256:[a-f0-9]{64}$'))`,message="the OCI Image reference must end with a valid '@sha256:<digest>' suffix, where '<digest>' is 64 characters long"
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@')[0].matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*?$') || self.matches('^[^.]+\\.[^.]+\\.svc:\\d+\\/[^\\/]+\\/[^\\/]+:[^\\/]+$'))`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme"
+	// +optional
+	FinalImagePullspec string `json:"finalImagePullspec"`
```
from an API consumer perspective, this is the `currentImagePullSpec`, right?

cdoern replied:
this is the version with the sha256. but, since what I think you're saying is that this is on the config level and can change, then yeah I can rename this current rather than final

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,233 @@
+
+// Refers to the name of an image registry push/pull secret needed in the build process.
+type ImageSecretObjectReference struct {
+	// name is the name of the secret used to push or pull this MachineOSConfig object.
```
which  namespace is this in?  having a hardcoded requirement makes sense.  We've generally used `openshift-config` for ones we don't own.

cdoern replied:
the secret itself? I believe we use two namespaces. one lives in MCO namespace, the other is created without a namespace which I am assuming defaults to some sane namespace (openshift-config). For the MCO one, we use the pull secret found in openshift-config and pull out data from it, then creating a copy in MCO...

I am open to changing this and putting the all in openshift-config, if so: how would I namespace this?

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,233 @@
+	// that only gives someone to pull images from the image repository. It's basically the principle of least permissions.
+	// this pull secret will be used on all nodes in the pool. These nodes will need to pull the final OS image and boot into it using rpm-ostree or bootc.
+	// +optional
+	RenderedImagePullSecret ImageSecretObjectReference `json:"renderedImagePullSecret"`
```
Why doesn't this option live on the MachineConfigPool?  Wouldn't this be natural to 
1. be able to configure even if I'm not using machineosconfig
2. be able to configure even if I cannot change the inputs to the built image
3. be able to update without triggering a new machineosbuild

cheesesashimi replied:
Yeah, I agree with @cdoern. While it might seem more natural to add these to a MachineConfigPool, it feels less natural when someone hasn't opted into on-cluster layering since these settings would be irrelevant in that situation. Because on-cluster layering is optional, I don't think it makes sense to directly attach this to the MachineConfigPools.

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,233 @@
+	// +kubebuilder:validation:XValidation:rule=`((self.split('@').size() == 2 && self.split('@')[1].matches('^sha256:[a-f0-9]{64}$')) || (self.split(':').size() == 2 && self.split(':')[1].matches('^([a-zA-Z0-9-./:])+$')) || self.matches('^[^.]+\\.[^.]+\\.svc:\\d+\\/[^\\/]+\\/[^\\/]+:[^\\/]+$'))`,message="the OCI Image reference must end with a valid '@sha256:<digest>' suffix or a valid :<tag>, wher ...
+	// +kubebuilder:validation:XValidation:rule=`((self.split('@').size() == 2 && self.split('@')[0].matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*?$')) || (self.split(':').size() == 2 && self.split(':')[0].matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*?$')) || self.matches('^[^.]+\\.[^.]+\\.svc:\\d+\\/ ...
+	// +kubebuilder:validation:Required
+	RenderedImagePullspec string `json:"renderedImagePullspec"`
```
from the perspective of the machineosconfig, this is an output, not an input.  it cannot SHA based since we don't know the sha in advance. And the spec is strictly a push spec, not a pull spec.

deads2k replied:
update to require being a tag.  Update to indicate it's a push spec.

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,233 @@
+	// +listMapKey=containerFileArch
+	// +optional
+	// +kubebuilder:validation:MinItems=0
+	// +kubebuilder:validation:MaxItems=2
```
with seven unumerated values, why only two items?

cdoern replied:
In a multi-arch setting, users should only be able to specify two architectures they want to build for.

deads2k replied:
During live review, agreed to make it match the number of arches.

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,233 @@
+	// +kubebuilder:validation:MinItems=0
+	// +kubebuilder:validation:MaxItems=2
+	// +optional
+	Containerfile []MachineOSContainerfile `json:"containerFile" patchStrategy:"merge" patchMergeKey:"containerFileArch"`
```
I may have misunderstood how the bits fit together, but aren't customizations supposed to be done in an machienconfig?

cheesesashimi replied:
The idea behind on-cluster layering is to give cluster admins more flexibility than they would have with MachineConfigs. I see the two models as complimentary to each other. The only major difference is how MachineConfigs get written to the nodes when on-cluster layering is enabled.

deads2k replied:
Request to add "this is a dockerfile" to the field doc.

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,233 @@
+	ContainerFileArch ContainerFileArch `json:"containerFileArch"`
+	// content is the custom content to be built
+	// +kubebuilder:validation:Required
+	Content string `json:"content"`
```
What if I wnt secret content?  If I cannot use this API (fine), how do I add secrets.  And if I can secrets using that mechanism, why shouldn't I add all content that way?

cheesesashimi replied:
I can see why someone may want to add build-time secrets, but with the exception of RHEL entitlements, we do not aim to provide that  functionality. The reason being that as we add things like that, we start to converge on features and workflows that would be better provided and implemented by a fully-blown CI / CD system.

That said, if someone wants to include content from, say, a private Git repository, they would need to build / push an image off-cluster containing that content and ensure that the image is pullable via whatever image pull secrets they provide.

deads2k replied:
Maybe I don't understand the content to be inject.  For instance, which file does this content end up in?

cdoern replied:
the content here is the content of the containerfile. We then build this into the final OS image. So for example if a user wants to install `cowsay` onto their nodes they specify it in the containerfile and then the cluster has that program built in.

This is less used to place content onto specific parts of the disk and more to augment the actual OS itself.

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,233 @@
+	// +kubebuilder:validation:Enum:=NoRebuild;RebuildNow;RebuildDelayed
+	// +kubebuilder:validation:Required
+	RebuildStrategy TriggerRebuild `json:"rebuildStrategy"`
+	// maxRetries is the amount of times we want to auto-rebuild after build failures before we deem the update unreconcilable.
```
why would we ever give up instead of simply having longer and longer timeouts until we're only retrying every few hours?

cdoern replied:
Is there a current methodology to do this? Or would this be handled on the controller side and here I would put  field called `timeout` or something.

cdoern commented:
A summary of changes I have made @deads2k 

in MachineOSBuild:
1. moved `ConfigGeneration` to spec, set minimum to 1
2. moved `Version` to spec, set minimum to 1
3. moved `DesiredConfig` to spec
4. added `ImageBuilderType` to status for the discriminated union
5. Created union in which `BuildPod` (was `Build`) is the only current member. But when we add more builder types, those will be added as members
6. renamed the pull spec to `finalImagePushSpec`
7. added an `ObjectReference` type that can generally track most API objects (build pods for example)

In MachineOSConfig
1. renamed `FinalImagePullspec` to `CurrentImagePullspec` as this can change through generations of the MachineOSConfig
2. Switched `MachineConfigPool` in spec to be `MachineOSConfig` as this is a more direct way of relating the MOSB to the MOSC. This should work in the MCO.
3. added back `ImageBuilderType` since we are going to be adding more backends in the future. The only enum for now is `PodImageBuilder`
4. Removed the `TriggerRebuild` and `MachineOSRebuild` types. Up in the air for what will replace these. A rebuild is triggered when the user changes the machineosconfig automatically. However, we want to allow users to *manually* trigger a rebuild. Though, this can wait for GA if need be.

Some clarifying points made in the comments:

OCL (On-Cluster-Layering) is a method of cluster configuration outside of the MachineConfig workflow. It coexists with MC and uses rendered configs as a point of reference so the image knows what to base the OS off of. However the `Containerfile` field allows the customer to layer content on top of the base rendered OS built from the MachineConfig. We want to keep this, as well as any references to image secrets, pullspecs, etc separate from MachineConfig and MachineConfigPool. We will support both workflows but they are intended for drastically different usecases, so we don't want to modify our existing objects with settings from the build process

Questions:

Did I do the discriminated union properly? the MCO is unfamiliar with this process.

A lot of your comments were based around "inputs vs outputs". I agree with all of them, however I want to make sure the user cannot set things like "ConfigGeneration" or "version" on the MOSB spec. I put them in status as that is commonly where the MCO puts entities that the controller itself owns. Do we risk a smooth workflow by putting these within the users reach? or can we enforce that the user cannot set these?

I moved around a bunch of stuff from status to spec so I may have left some extra questions in the comments above, please let me know if anything is unclear!

hexfusion commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,199 @@
+type MachineOSConfigSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineConfigPoolReference `json:"machineConfigPool"`
```
Why couldn't a build target more than one pool?

cdoern replied:
this is not in scope for OCL. one config per pool and then multiple build objects per config. Muddying the waters here and opening up the door for 3+ pools to use the same config will make implementation and UX much worse I think

hexfusion replied:
OK so its not possible for an OS image to be generic in this context is what your saying.

cdoern replied:
in this scope, yes.

cdoern replied:
and honestly, in the scope of OCL for the foreseeable future. MCO is only in the business of managing images we are going to apply for you not images that just hang out in a registry.

hexfusion commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,199 @@
+// MachineOSConfig describes the configuration for a build process managed by the MCO
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineOSConfig struct {
```
I keep reading this and it feels like it should be `MachineOsBuildConfig`?

cdoern replied:
ehhh, I think MachineOSConfig is good. it is the config for building the OS, we will probably over time add more options here that lean towards configuring the OS rather than the build process, naming it `MachinOSBuildConfig` will close that door and create the opposite issue. 

WDYT @cheesesashimi ?

hexfusion replied:
Ok but logically you have two separate things here or maybe I  am misunderstanding.
- a build with a config
- an os image which is part of the build which also has a config

hexfusion replied:
So given above would the build process itself ever need its own configurations? In my mind its like this.
```
build:
  - config:
    - image:
       - name: 
       - config: 
    - other_build_params:
```

cdoern replied:
your thinking of the structure makes sense to me,  I was just saying I'd like to keep the name open, given that we have a roadmap of different spec entities we are going to add to MachineOSConfig. I don't want to tie it too much to the name "Build". That is what the `MachineOSBuild` is for :)

hexfusion commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,199 @@
+	// +kubebuilder:validation:MaxLength=447
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@').size() == 2 && self.split('@')[1].matches('^sha256:[a-f0-9]{64}$'))`,message="the OCI Image reference must end with a valid '@sha256:<digest>' suffix, where '<digest>' is 64 characters long"
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@')[0].matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*?$'))`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme"
+	// +optional
```
I guess same question here about optional might be worth adding tests for status I added some recently to MCN which helped me validate these are doing what you want.

cdoern replied:
fixed

cdoern replied:
you were right, these were causing issues in MCO when unset and I wasn't sure why

deads2k commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,155 @@
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// ImageBuilderType describes the image builder set in the MachineOSConfig
+	// +unionDiscriminator
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+	// relatedObjects is a list of objects that are related to the build process.
+	// +unionMember,optional
+	BuildPod []ObjectReference `json:"buildPod,omitempty"`
```
```go
type MachineOSBuildStatus struct {
	BuilderReference *MachineOSBuilderReference
}

type MachineOSBuilderReference struct{
	// ImageBuilderType describes the image builder set in the MachineOSConfig
	// +unionDiscriminator
	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`

	// relatedObjects is a list of objects that are related to the build process.
	// +unionMember,optional
	PodImageBuilder []ObjectReference `json:"buildPod,omitempty"`
}



```

deads2k commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,155 @@
+	// ImageBuilderType describes the image builder set in the MachineOSConfig
+	// +unionDiscriminator
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+	// relatedObjects is a list of objects that are related to the build process.
```
```go
type MachineOSBuildStatus struct {
	RelatedObjects []ObjectReference `json:"relatedObjects,omitempty"`
}
```

you can put things like
1. namespaces
2. direct configmaps (I was configured from here)
3. MachineOSConfig
4. MachineConfigPool

deads2k commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+// MachineOSBuildSpec describes information about a build process primarily populated from a MachineOSConfig object.
+type MachineOSBuildSpec struct {
```
we need to avoid this

1. machineosconfig rv=1 this is secure, pushes to location-secure
2. machineosbuild/A is created for this secure image
3. machineosconfig rv=2 is insecure, pushes to location-unsafe
4. new machineosbuild for whatever reason is not created
5. machineosbuild/A is built and looks at machineosconfig sees to push to location-unsafe, exposes secrets.

to avoid this, add an imagepushspec that must be a tag for where to push the resulting image.

deads2k replied:
All this content is known on create, therefore we should not allow any update to this information.  You can enforce this using CEL + oldSelf

deads2k commented on machineconfiguration/v1alpha1/types_machineosbuild.go:
```diff
@@ -0,0 +1,155 @@
+// +kubebuilder:validation:XValidation:rule="has(self.imageBuilderType) && self.imageBuilderType == 'PodImageBuilder' ?  true : !has(self.buildPod)",message="buildPod is required when imageBuilderType is PodImageBuilder, and forbidden otherwise"
+type MachineOSBuildStatus struct {
+	// conditions are state related conditions for the build. Valid types are:
+	// Prepared, Building, Failed, Interrupted, Restarted, and Ready
```
s/Ready/Succeeded/g

Restarted is gone, right?

Terminal states are Failed and Succeeded.

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,199 @@
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@')[0].matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*?$'))`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme"
+	// +optional
+	BaseOSImagePullspec string `json:"baseOSImagePullspec"`
+	// baseImagePullSecret is the secret used to pull the base image.
```
specify which namespace this must be in.  MCO namespace is fine, but be very specific so people can consume this well.

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,199 @@
+	// +kubebuilder:default:=PodImageBuilder
+	// +kubebuilder:validation:Enum:=PodImageBuilder
+	// Valid options are: PodImageBuilder
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
```
I bet this ends up needing to be a discriminated union.

```go

type BuildSettings struct{
  ImageBuilder *MachineOSImageBuilder
}

type MachineOSImageBuilder struct{
  ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
}

```

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,199 @@
+	// that only gives someone to pull images from the image repository. It's basically the principle of least permissions.
+	// this pull secret will be used on all nodes in the pool. These nodes will need to pull the final OS image and boot into it using rpm-ostree or bootc.
+	// +optional
+	RenderedImagePullSecret ImageSecretObjectReference `json:"renderedImagePullSecret,omitempty"`
```
`CurrentImagePullspecSecret`

deads2k replied:
This one isn't a BuildSetting.  Perhaps moving to a different area in spec?

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,199 @@
+}
+
+// MachineOSConfigStatus describes the status this config object and relates it to the builds associated with this MachineOSConfig
+type MachineOSConfigStatus struct {
```
Recommend adding Conditions.

cdoern commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,199 @@
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=447
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@').size() == 2 && self.split('@')[1].matches('^sha256:[a-f0-9]{64}$'))`,message="the OCI Image reference must end with a valid '@sha256:<digest>' suffix, where '<digest>' is 64 characters long"
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@')[0].matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*?$'))`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme"
```
fixed

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,224 @@
+	// +kubebuilder:validation:Required
+	RenderedImagePushspec string `json:"renderedImagePushspec"`
+	// releaseVersion is associated with the base OS Image. This is the version of Openshift that the Base Image is associated with.
+	// +optional
```
What happens when this isn' specified?  How is this validated?  What format shoudl it be in?  how wil it work pre-release?  What is it used for?

cdoern replied:
added godoc

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,224 @@
+	// +optional
+	BaseOSImagePullspec string `json:"baseOSImagePullspec,omitempty"`
+	// baseImagePullSecret is the secret used to pull the base image.
+	// must live in the MCO namespace
```
throughout the doc, specify the actual namespace.  openshift-machine-config-somthing?

cdoern replied:
`openshift-machine-config-operator` and fixed

deads2k commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,228 @@
+	// releaseVersion is associated with the base OS Image. This is the version of Openshift that the Base Image is associated with.
+	// This field is populated from the machine-config-osimageurl configmap in the openshift-machine-config-operator namespace if not user specified.
+	// If the user chooses to specify, it should come in the format: 4.16.0-0.nightly-2024-04-03-065948 or any valid release. The MachineOSBuilder validates that this is a valid stream.
+	// If overwirtten after already set, the MachineOSBuilder will kick off a build with the new setting
```
does the user own this field or the controller moving it from teh configmap. They cannot share.

deads2k commented:
/approve

cheesesashimi reviewed (commented):
Almost there! Just a couple very minor things left.

cdoern commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,227 @@
+	// +kubebuilder:validation:Enum:=arm64;amd64;ppc64le;s390x;aarch64;x86_64;noArch
+	// +kubebuilder:default:=noArch
+	// +optional
+	ContainerFileArch ContainerFileArch `json:"containerFileArch"`
```
fixed

cdoern commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,227 @@
+	Aarch64 ContainerFileArch = "aarch64"
+	// describes the x86_64 architecture
+	X86_64 ContainerFileArch = "x86_64"
+	// describes a containerfile that can be applied to any arch
```
I think this can be added later, or maybe even in the OCP docs, not these.

cdoern commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -0,0 +1,227 @@
+	// describes the x86_64 architecture
+	X86_64 ContainerFileArch = "x86_64"
+	// describes a containerfile that can be applied to any arch
+	NoArch ContainerFileArch = "noArch"
```
fixed

cheesesashimi commented:
/lgtm
/approve

cheesesashimi commented:
/lgtm
/approve
